go-controller/pkg/ovn/namespace.go

```diff
@@ -218,7 +236,66 @@ func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
 	}
 	defer nsInfo.Unlock()
 
-	annotation := newer.Annotations[hotypes.HybridOverlayExternalGw]
+	var annotation, oldAnnotation string
+	annotation = newer.Annotations[routingExternalGWsAnnotation]
+	oldAnnotation = old.Annotations[routingExternalGWsAnnotation]
+	if annotation != oldAnnotation {
+		var stderr string
+		var err error
+		for podIP, gwToGr := range nsInfo.podExternalRoutes {
+			for gw, gr := range gwToGr {
+				// TODO figure out how to delete route with gw
+				_, stderr, err = util.RunOVNNbctl("--if-exists", "--policy=src-ip",
+					"lr-route-del", gr, podIP+"/32", gw)
+			}
+			if err != nil {
+				klog.Errorf("Unable to delete src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+			} else {
+				delete(nsInfo.podExternalRoutes, podIP)
+			}
+		}
+		if annotation != "" {
+			annotationOk := true
+			nsInfo.routingExternalGWs = nil
+			for _, v := range strings.Split(annotation, ",") {
+				parsedAnnotation := net.ParseIP(v)
+				if parsedAnnotation == nil {
+					klog.Errorf("Could not parse routing external gw annotation")
+					annotationOk = false
+					nsInfo.routingExternalGWs = nil
+					break
+				} else {
+					nsInfo.routingExternalGWs = append(nsInfo.routingExternalGWs, parsedAnnotation)
+				}
+			}
+			if annotationOk {
+				existingPods, err := oc.watchFactory.GetPods(old.Name)
+				if err != nil {
+					klog.Errorf("Failed to get all the pods (%v)", err)
+				} else {
+					for _, pod := range existingPods {
+						podIP := pod.Status.PodIP
+						gr := "GR_" + pod.Spec.NodeName
+						for _, gw := range nsInfo.routingExternalGWs {
+							_, stderr, err = util.RunOVNNbctl("--may-exist", "--policy=src-ip", "--ecmp",
+								"lr-route-add", gr, podIP+"/32", gw.String())
+							if err != nil {
+								klog.Errorf("Unable to add src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+							} else {
+								if nsInfo.podExternalRoutes[podIP] == nil {
+									nsInfo.podExternalRoutes[podIP] = make(map[string]string)
+								}
+								nsInfo.podExternalRoutes[podIP][gw.String()] = gr
+							}
+						}
+					}
+				}
+			} else {
+				nsInfo.routingExternalGWs = nil
+			}
+		}
```

Comment:
it looks like nothing clears `nsInfo.routingExternalGWs` if you unset the annotation

Reply:
ack

Reply:
done